apps/ewallet_db/priv/repo/migrations/20190510071247_add_two_fa_to_user.exs

```diff
@@ -0,0 +1,13 @@
+defmodule EWalletDB.Repo.Migrations.AddTwoFAToUser do
+  use Ecto.Migration
+
+  def change do
+    alter table(:user) do
+      add :enabled_2fa_at, :naive_datetime
+      add :secret_2fa_code, :string
+      add(:hashed_backup_codes, {:array, :string}, default: [])
+      add :used_backup_code_at, :naive_datetime
+      add(:used_hashed_backup_codes, {:array, :string}, default: [])
```

Comment:
PostgreSQL array can get slow really fast, and the way to manipulate them by _updating_ the field may subject to race condition without proper locking (e.g. when user try to use two backup codes at the same time from different session).

I think it's better to follow [1NF](https://en.wikipedia.org/wiki/First_normal_form) design and have a `user_backup_code` table that store an ID of a user, a hashed backup code, a date of creation, and a date of use (in which setting date of use invalidated the code). 

Sorry, I just caught this.

Reply:
👍 Will fix it

Reply:
@sirn Done krub
Implementation: https://github.com/omisego/ewallet/pull/920/commits/926849a130f5a420496d042ffe97576eeed83673
Fixed tests and add some:
https://github.com/omisego/ewallet/pull/920/commits/8ce8b385621fd88dc1104cf67a7d433227ec55c6

Reply:
@sirn Please re-review 🙏 

https://github.com/omisego/ewallet/pull/920/files/894af2be4206e0e8d68d9469c8a368196a530e7f..52e97a72157c13f8bc0ff0ad988f229e2b48ef66